dask_pytorch/data.py

```diff
@@ -0,0 +1,97 @@
+"""
+Utilities for loading PyTorch data in distributed environments
+"""
+
+
+import tempfile
+from os.path import basename, dirname
+
+from PIL import Image
+from torch.utils.data import Dataset
+
+
+"""
+In the following, we are explicitly avoiding s3fs because it does not behave well with
+multiprocessing (which is commonly used in PyTorch dataloaders).
+
+https://github.com/dask/s3fs/issues/369
+"""  # pylint: disable=pointless-string-statement
+
+
+def _list_all_files(bucket: str, prefix: str, s3_client=None) -> List[str]:
+    """
+    Get list of all files from an s3 bucket matching a certain prefix
+    """
+    import boto3  # pylint: disable=import-outside-toplevel
+
+    if s3_client is None:
+        s3_client = boto3.client("s3")
+    paginator = s3_client.get_paginator("list_objects")
+    all_files = []
+    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
+        files = [x["Key"] for x in page["Contents"]]
+        all_files.extend(files)
+    return all_files
+
+
+def _read_s3_fileobj(bucket, path, fileobj):
+    """
+    read an obj from s3 to a file like object
+    """
+    import boto3  # pylint: disable=import-outside-toplevel
+
+    s3 = boto3.resource("s3")
+    bucket = s3.Bucket(bucket)
+    bucket.download_fileobj(path, fileobj)
+    fileobj.seek(0)
+    return fileobj
+
+
+def _load_image_obj(fileobj):
+    """
+    turn a file like object into an image
+    """
+    return Image.open(fileobj).convert("RGB")
+
+
+class S3ImageFolder(Dataset):
+    """
+    An image folder that lives in S3.  Directories containing the image are classes.
+    """
+
+    def __init__(self, s3_bucket: str, s3_prefix: str, transform: Optional[Callable] = None, target_transform: Optional[Callable] = None):
+        self.s3_bucket = s3_bucket
+        self.s3_prefix = s3_prefix
+        self.all_files = _list_all_files(s3_bucket, s3_prefix)
+        self.classes = sorted({self.get_class(x) for x in self.all_files})
+        self.class_to_idx = {k: idx for idx, k in enumerate(self.classes)}
+        self.transform = transform
+        self.target_transform = target_transform
+
+    @classmethod
+    def get_class(cls, path):
```

Comment:
```suggestion
    def _get_class(cls, path):
```

This looks like it could be internal too, I think.

Reply:
A private method requires double underscores preceding, right?

Reply:
oh yeah, you're absolutely right.

I usually do a single leading underscore which is like "ok I'm telling you to please not access this, but if you want to you technically can, just know that I get to break it any time I want".

The double leading underscore is harsher and is like "you can never call this".

```python
class SomeClass:
    def _something_private(self):
         return("a")
    def __something_very_private(self):
        return("b") 

# try it
sc = SomeClass()

# no error
sc._something_private()

# error
sc.__something_very_private()
```

I lean towards single-underscore for libraries and double-underscore for applications, but it's not a preference I hold strongly. So whichever you want!

Reply:
There are other private methods on the class with double underscores so I'll stick with that standard here.

Reply:
> There are other private methods on the class with double underscores so I'll stick with that standard here.

I think the only ones we have are `__getitem__`?  That's a bit different.  Single is more common but I don't have a good reason